### PR TITLE
Memory Profiling

### PR DESCRIPTION
**Pull request by madsbk**
Use [RMM's new memory profiler](https://github.com/rapidsai/rmm/pull/1563) to profile all functions already decorated with `_cudf_nvtx_annotate`.

Example 
```python
import cudf
from cudf.utils.performance_tracking import print_memory_report

cudf.set_option("memory_profiling", True)

df1 = cudf.DataFrame({"a": [1, 2, 3]})
df2 = cudf.DataFrame({"a": [2, 2, 3]})
df3 = df1.merge(df2)

print_memory_report()
```

Output:
```
Memory Profiling
================

Ordered by: memory_peak

ncalls     memory_peak    memory_total  filename:lineno(function)
     1             272             688  /home/mkristensen/apps/miniforge3/envs/rmm-cudf-0527/lib/python3.11/site-packages/cudf/core/dataframe.py:4072(DataFrame.merge)
     2              32              64  /home/mkristensen/apps/miniforge3/envs/rmm-cudf-0527/lib/python3.11/site-packages/cudf/core/dataframe.py:1043(DataFrame._init_from_dict_like)
     2              32              64  /home/mkristensen/apps/miniforge3/envs/rmm-cudf-0527/lib/python3.11/site-packages/cudf/core/dataframe.py:690(DataFrame.__init__)
     2               0               0  /home/mkristensen/apps/miniforge3/envs/rmm-cudf-0527/lib/python3.11/site-packages/cudf/core/dataframe.py:1131(DataFrame._align_input_series_indices)
     7               0               0  /home/mkristensen/apps/miniforge3/envs/rmm-cudf-0527/lib/python3.11/site-packages/cudf/core/index.py:214(RangeIndex.__init__)
     6               0               0  /home/mkristensen/apps/miniforge3/envs/rmm-cudf-0527/lib/python3.11/site-packages/cudf/core/index.py:424(RangeIndex.__len__)
     4               0               0  /home/mkristensen/apps/miniforge3/envs/rmm-cudf-0527/lib/python3.11/site-packages/cudf/core/frame.py:271(Frame.__len__)
     2               0               0  /home/mkristensen/apps/miniforge3/envs/rmm-cudf-0527/lib/python3.11/site-packages/cudf/core/dataframe.py:3195(DataFrame._insert)
     2               0               0  /home/mkristensen/apps/miniforge3/envs/rmm-cudf-0527/lib/python3.11/site-packages/cudf/core/index.py:270(RangeIndex.name)
     2               0               0  /home/mkristensen/apps/miniforge3/envs/rmm-cudf-0527/lib/python3.11/site-packages/cudf/core/index.py:369(RangeIndex.copy)
     5               0               0  /home/mkristensen/apps/miniforge3/envs/rmm-cudf-0527/lib/python3.11/site-packages/cudf/core/frame.py:134(Frame._from_data)
     2               0               0  /home/mkristensen/apps/miniforge3/envs/rmm-cudf-0527/lib/python3.11/site-packages/cudf/core/frame.py:1039(Frame._copy_type_metadata)
     2               0               0  /home/mkristensen/apps/miniforge3/envs/rmm-cudf-0527/lib/python3.11/site-packages/cudf/core/indexed_frame.py:315(IndexedFrame._from_columns_like_self)
```

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [x] Rename `_cudf_nvtx_annotate` to `_performance_tracking`

### CODE REVIEW

File: python/cudf/cudf/utils/nvtx_annotation.py
Comment (madsbk) [inline python/cudf/cudf/utils/nvtx_annotation.py:21]: Should we rename `_cudf_nvtx_annotate` to something like `_performance_tracking` ?

Comment (mroeschke) [reply python/cudf/cudf/utils/nvtx_annotation.py:21]: I think that would be a more generically clear name. +1

**Comment by harrism** [comment]
I requested this in https://github.com/rapidsai/rmm/pull/1563, but it seems not to have made it in. The current output shown above provides no way of knowing what the units are in the peak and total memory report. Suggest changing the headings to make this explicit, otherwise usability is reduced.

```
ncalls     memory_peak (MiB)    memory_total (MiB)  filename:lineno(function)
...
```

**Comment by madsbk** [comment]
@harrism, I added a Legends section

```
Memory Profiling
================

Legends:
  ncalls       - number of times the function or code block was called
  memory_peak  - peak memory allocated in function or code block (in bytes)
  memory_total - total memory allocated in function or code block (in bytes)
```

**Comment by bdice** [comment]
@madsbk Is there any measurable overhead to using the statistic profiler? The NVTX overhead is very small, which is why we have enabled it across cuDF. We may want to run benchmarks to verify this before merging. If it has a cost, perhaps use an environment variable such as `CUDF_PROFILE_MEMORY_STATISTICS` to control the behavior.

**Comment by madsbk** [comment]
Good point, it has a price of 1.7 microseconds.
The overhead goes from 0.18us to 1.9us. The extra overhead is dominated by two calls to `rmm.mr.get_current_device_resource()`, which takes 1.5us.

I think this is small enough to ignore?

**Comment by bdice** [comment]
> The overhead goes from 0.18us to 1.9us.

@vyasr @galipremsagar Do you think this is small enough to enable memory statistics tracking on every NVTX-annotated function in cudf Python? I am pretty hesitant to enable this by default but don't want to be the only person weighing in.

**Comment by harrism** [comment]
Given the overhead is constant, when memory size is high, this will be a drop in the bucket. When it's low, it may be significant on some functions. 

Can we just make it really easy to enable, rather than enabled by default?

My strong opinion, weakly held is that nonzero cost profiling should not be enabled by default.

**Comment by madsbk** [comment]
Alright, I have added a `memory_profiling` option :)


It also makes it a bit more user-friendly. To enable memory profiling do: 
```python
import cudf
from cudf.utils.performance_tracking import print_memory_report

cudf.set_option("memory_profiling", True)

# my code

print_memory_report()
```


**NB**: before merging this PR, I will remove `nvtx_annotation.py` and use `@_performance_tracking` throughout.

File: python/cudf/cudf/utils/performance_tracking.py
Comment (wence-) [inline python/cudf/cudf/utils/performance_tracking.py:73]:
```suggestion
    """Pretty print the result of the memory profiling
```

File: python/cudf/cudf/utils/utils.py
Comment (wence-) [inline python/cudf/cudf/utils/utils.py:164]: What has changed such that this is necessary?

File: python/cudf/cudf/utils/performance_tracking.py
Comment (wence-) [inline python/cudf/cudf/utils/performance_tracking.py:33]: *issue*: This doesn't behave like a normal context manager (leave the state unchanged after exit) because it doesn't unwind the change to the current memory resource that `rmm.statistics.enable_statistics` enacts. This means that the following code does an "unexpected" thing: side-effectfully changes the MR:

```python
# statistics not enabled...
with cudf.options_context(("memory_profiling", True)):
    # code that profiles statistics

# expect stats memory resource to be popped, but it is not
```

It seems like the call to enable statistics should be paired with a matching disable statistics call. But this seems bad if it were done at every level of the call stack.

It _feels_ like the right place for modification of the MR is in the `set_option` call. `set_option("memory_profiling", True)` would store the old MR, push the statistics MR, and then a later `set_option("memory_profiling", False")` would pop the statistics MR if it still matches.

This, however, would preclude use of the environment variable default (because that is set on import, and we don't want to have to call RMM on import).

Does this suggest that we need a specific `enable_memory_profiling` that can be used as a context manager that sets the option and sets up the MR?

**Review by wence- — CHANGES_REQUESTED** [review]
I think we want to figure out a way to be able to disable the statistics MR, which is currently not possible.

File: python/cudf/cudf/utils/utils.py
Comment (madsbk) [reply python/cudf/cudf/utils/utils.py:164]: I am not sure, but I think it is because we now use a context always. 

For some reason, [`walk_stack()`](https://github.com/python/cpython/blob/main/Lib/traceback.py#L383) and [`extract_stack()`](https://github.com/python/cpython/blob/main/Lib/traceback.py#L383) use different default start frames.

File: python/cudf/cudf/utils/performance_tracking.py
Comment (madsbk) [reply python/cudf/cudf/utils/performance_tracking.py:33]: If this is an issue, RMM already have `rmm.statistics.statistics()` and `rmm.statistics.enable_statistics()`.
Thus the user code would look something like:
```python
import cudf
import rmm.statistics

# The must first enable the cudf option
cudf.set_option("memory_profiling", True)

# Then enable statistics within a context
with rmm.statistics.statistics():
    # by code

# Or globally
rmm.statistics.enable_statistics()
```
Of cause, we could wrap/alias the rmm functions in cudf native functions. However, I am not sure if the disadvantage of the "unexpected" side-effect outweigh the user-friendliness of just setting `CUDF_MEMORY_PROFILING="1"`?

Comment (madsbk) [reply python/cudf/cudf/utils/performance_tracking.py:33]: Alternatively, we could implement lazy environment variable in `options.py`. It might be worth streamline the option module in any case.

Comment (wence-) [reply python/cudf/cudf/utils/performance_tracking.py:33]: In terms of solutions, it depends a little bit on how end-user-friendly we want this to be. If this is mostly for developer use, I am happy with the programmatic set_option + use RMM calls. If we want this as a debugging tool for end users "where is my code allocating lots of vram", then we likely need something a little more ergonomic.

As much as I am allergic to environment variables from a reproducibility point of view, I can see the attraction.

Comment (madsbk) [reply python/cudf/cudf/utils/performance_tracking.py:33]: I am not sure but I guess we could start with the `set_option + use RMM calls` and then later consider a more end-user-friendly solution?

Comment (wence-) [reply python/cudf/cudf/utils/performance_tracking.py]: I think that's a reasonable compromise.

Comment (madsbk) [reply python/cudf/cudf/utils/performance_tracking.py]: done

File: docs/cudf/source/user_guide/memory-profiling.md
Comment (harrism) [inline docs/cudf/source/user_guide/memory-profiling.md:5]:
```suggestion
Peak memory usage is a common concern in GPU programming because GPU memory is typically smaller than available CPU memory. To easily identify memory hotspots, cuDF provides a memory profiler. It comes with an overhead so avoid using it in performance-sensitive code.
```

File: docs/cudf/source/user_guide/memory-profiling.md
Comment (harrism) [inline docs/cudf/source/user_guide/memory-profiling.md:11]: This is confusing because you have "First..." and "Second", but also "One way to do this..." in between these. Then you have two numbered bullets. So there are too many numbered things in this section. Can you simplify it? 

Even better, why do we have to enable memory profiling in RMM at all -- can't cuDF handle that for us if we enable it in cuDF?

**Review by harrism — CHANGES_REQUESTED** [review]
Doc comments.

File: docs/cudf/source/user_guide/memory-profiling.md
Comment (madsbk) [reply docs/cudf/source/user_guide/memory-profiling.md]: Tried to simplify it a bit. 

> Even better, why do we have to enable memory profiling in RMM at all -- can't cuDF handle that for us if we enable it in cuDF?

I did this initially but as @wence- points out, by enabling RMM statistics implicitly: [This doesn't behave like a normal context manager (leave the state unchanged after exit)](https://github.com/rapidsai/cudf/pull/15866#discussion_r1650602241)

Comment (harrism) [reply docs/cudf/source/user_guide/memory-profiling.md]: OK got it.

File: docs/cudf/source/user_guide/memory-profiling.md
Comment (harrism) [inline docs/cudf/source/user_guide/memory-profiling.md:11]: Suggestion: better section title, remove "royal we" and "one way", and link to the specific RMM docs page (link won't work until 24.08 is released, but I don't want to link to nightly docs).

```suggestion
## Enabling Memory Profiling

First, enable memory profiling in RMM by calling {py:func}`rmm.statistics.enable_statistics()`. This adds a statistics resource adaptor to the current RMM memory resource, which enables cudf to access memory profiling information. See the [RMM documentation](https://docs.rapids.ai/api/rmm/stable/guide/#memory-statistics-and-profiling) for more details.

Second, enable memory profiling in cudf by setting the `memory_profiling` option to `True`. Use {py:func}`cudf.set_option` or set the environment variable ``CUDF_MEMORY_PROFILING=1`` prior to the launch of the Python interpreter.
```

**Review by harrism — APPROVED** [review]

File: docs/cudf/source/user_guide/memory-profiling.md
Comment (wence-) [inline docs/cudf/source/user_guide/memory-profiling.md:9]:
```suggestion
First, enable memory profiling in RMM by calling {py:func}`rmm.statistics.enable_statistics()`. This adds a statistics resource adaptor to the current RMM memory resource, which enables cuDF to access memory profiling information. See the [RMM documentation](https://docs.rapids.ai/api/rmm/stable/guide/#memory-statistics-and-profiling) for more details.
```

File: docs/cudf/source/user_guide/memory-profiling.md
Comment (wence-) [inline docs/cudf/source/user_guide/memory-profiling.md:11]:
```suggestion
Second, enable memory profiling in cuDF by setting the `memory_profiling` option to `True`. Use {py:func}`cudf.set_option` or set the environment variable ``CUDF_MEMORY_PROFILING=1`` prior to the launch of the Python interpreter.
```

**Review by wence- — APPROVED** [review]
Looks good, tiny typographic suggestions!

File: python/cudf/cudf/utils/nvtx_annotation.py
Comment (vyasr) [inline python/cudf/cudf/utils/nvtx_annotation.py:9]: Is this removable now?

**Review by vyasr — CHANGES_REQUESTED** [review]
One question, otherwise LGTM.

File: python/cudf/cudf/utils/nvtx_annotation.py
Comment (wence-) [reply python/cudf/cudf/utils/nvtx_annotation.py:9]: @madsbk was waiting for approvals before making this change due to it likely conflicting with _everything_, and plans to do this as the last step before merge

Comment (vyasr) [reply python/cudf/cudf/utils/nvtx_annotation.py:9]: Got it. Approving, then.

**Review by vyasr — APPROVED** [review]

**Review by wence- — APPROVED** [review]
Approving dask-cudf changes

**Comment by madsbk** [comment]
Thanks all
/merge

**Pull request merged**
